Define EMSCRIPTEN in the wasm backend

kripken opened the pull request:
This is useful for fastcomp backwards compatibility.

This is in addition to the more proper `__EMSCRIPTEN__` that the backend defines by itself.

Alternatively, perhaps we can not do this, and document in a prominent place that this is a breaking change.

Fixes #8921

VirtualTim commented:
There is a comment about this hidden deep in the code, so a more obvious place would be good.

https://github.com/emscripten-core/emscripten/blob/9aff7f29885ed34ea42f129ec718f3c0f0fdd018/emcc.py#L672-L673

sbc100 commented:
Yeah, it looks like it should be defined unless you build with `-s STRICT`

kripken commented:
Thanks @VirtualTim and @sbc100, very good points. Fixed in that commit.

We actually did have code to do this, but it was just wrong (it added the define in the wrong place) and had no test. Fixed it and added testing.

VirtualTim commented on src/settings.js:
```diff
@@ -746,6 +746,11 @@ var LINKABLE = 0;
+//   * ERROR_ON_MISSING_LIBRARIES is enabled
+//   * DISABLE_DEPRECATED_FIND_EVENT_TARGET_BEHAVIOR is enabled
+//   * The C define EMSCRIPTEN is not defined (__EMSCRIPTEN__ always is, and
+//     is the correct thing to use).
```
👍 This is much more obvious.

sbc100 commented on emcc.py:
```diff
@@ -1082,10 +1086,6 @@ def check(input_file):
-    else:
-      # The preprocessor define EMSCRIPTEN is deprecated. Don't pass it to code
-      # in strict mode. Code should use the define __EMSCRIPTEN__ instead.
-      shared.COMPILER_OPTS += ['-DEMSCRIPTEN']
```
Why is this location wrong?

kripken replied:
It was wrong because `shared.COMPILER_OPTS` is read and added to `newargs`. Modifying it after that point is never noticed.
